### DISCUSSION TITLE
YS-WK4T 4-channel relay with 5 inputs and 2 buttons

### DISCUSSION THREAD

**Discussion by thigger** [arendst/Tasmota#15822] [Show and tell]
Thought I should post something as I picked up one of these modules and I can't see anything about using it with Tasmota.

![_-9KvKsuSm29lMd3BuMoUg](https://user-images.githubusercontent.com/10333337/174445892-ebc9fd58-382c-43f5-8761-6d0b69bcbe7c.jpg)

Module is labelled YS-WK4T. On board has a PSF-B04 and a STC15W408AS.
Four relays, two physical buttons (K1, K2) and five inputs (K1-5 confusingly) connected via optocouplers - activated on connection to the board's 5V

Flashing - the headers on the board don't give RX or TX to the PSF-B04, but do have 3v3 and GND. I connected power there and RX/TX/GPIO0 directly on the PSF-B04 module.

In Tasmota, the module seems to respond correctly using the "Sonoff 4CH Pro" settings, except for the following behaviours.

Inputs K1-4 map to Buttons 1-4, though a connection to 5V is mapped to a 1-0 and then 0-1 transition 200ms later. Going back to floating does nothing.

Button K1 maps to Button1 "live" - ie 1-0 transition when pressed, 0-1 when let go

Button K2 acts as an "all off" button - does nothing unless relays are on, in which case it simulates a 1-0 then 
0-1 transition (after approx 200ms) sequentially through any button associated with a relay which is on. 

Input K5 acts as an "all on/off" button by simulating button 1-0 and then 0-1 (200ms gap) for buttons associated with relays which need their state changing. It's not clear to me how it determines whether to go all on, or all off.

Hope this is useful to someone! It's not quite clear to me yet why exactly the STC is in there, but it seems to all be working-ish (frustrating that K1-4 are only momentary)

**Comment by Pewibe** [comment]
Yes, this might be useful for me because I'm looking for a 4..8 ch relay switch 230VAC / relay input 230VAC DIN rail mountable for years. Since begin of this year Shelly pro 4pm might be the ultimative device if there were a more competitive price than 80..90 €. 
So may I ask for more information:

1. Buying source, manufacturers web site?
2. Datasheet, dimensions of the case, DIN rail?
3. Only external antenna or onboard alternative?
4. Type of opto coupler? Chance for 230VAC instead of 5VDC?

Thanks in advance for your additional info.

**Comment by thigger** [comment]
I bought it from aliexpress ( https://www.aliexpress.com/item/32789559654.html ).
DIN rail, case is 115mm x 88mm x 40mm. No datasheet that I could find online, or with it (there was a QR code to download the Ewelink app but I never tested it before installing Tasmota. I do have a backup of the original ROM if anyone wants though).
The ceramic antenna is still on the PS4-B04 module, it looks like they're connected in parallel - I have no idea how this affects transmission. Certainly I haven't put the antenna on yet for testing and it works fine.
Optocouplers are UMW 817C 2202 - doesn't look like they're 220V tolerant.

**Reply by sfromis** [reply]
According to the Itead "[Application Guide](https://itead.cc/diy-kits-guides/psf-b04-application-guide/)" for PSF-B04, it would appear that the designated antenna pin 1 "cannot be used", which would imply that only the ceramic antenna is used. I cannot say for certain if this PSF-B04 could be a "slightly different clone"....

**Reply by thigger** [reply]
They've put a bit of solder from the ceramic antenna back to that pin, so the external antenna is definitely connected. The ceramic one is pretty good though - I was still able to control it from the end of my garden where my phone has normally lost connection. Adding the external antenna back in improved the connection further - I'm planning to use a cable to have the antenna external to the main enclosure.

![OQk_qxQCSy6OODRLoQ3yDQ](https://user-images.githubusercontent.com/10333337/174474298-704cc7f3-96c5-47cb-8aa5-b5d494ff98a6.jpg)

**Reply by sfromis** [reply]
So that confirms the application guide warning....

I do not pretend to know much about the very tricky considerations around designing antenna circuits, but seeing how modules with U.FL sockets for external antenna generally have a small jumper (zero-ohm "resistor") to be connected to *either* the built-in antenna (like a PCB trace antenna) *or* the external antenna, I'm "unimpressed" by that bodge.

In many cases, the available benefit with external antennas is access to better antenna placement than inside some box, or "shielded" by nearby parts. And that PCB design is outright horrible around how the ceramic antenna is not given as much metal-free PCB space as possible, but directly placed away from the edge.

**Comment by Pewibe** [comment]
It looks like it's the only 4 channel device in Aliexpress or even other shops equipped with input terminals. @thigger What was your main reason for buying it? Just technical curious? The inputs? I guess it's not the price.

**Reply by thigger** [reply]
Input terminals and the ready-made ability to connect an external antenna for use in an enclosure were the main two reasons. It's installed now and works pretty well for my purposes so far (it's replacing a homebrew device I rigged up using 433MHz and an adaptation of the LightwaveRF protocol, which sometimes had spotty reception).
Main irritation is that the inputs are only detectable on a rising edge; though I've managed to rig up two inputs with a rising edge on each edge of the real signal, and a Tasmota rule to turn it back into an OPEN/CLOSED contact.

**Reply by Pewibe** [reply]
Ok, understood and your case makes sense. So I continue to wait for my dream device for upgrade of my DIN rail batteries of Eltako relays. :-)
